src/scipp/signal/__init__.py

```diff
@@ -0,0 +1,105 @@
+    """
+    Butterworth digital and analog filter design.
+
+    Design an Nth-order digital or analog Butterworth filter and return the filter
+    coefficients.
+
+    This is intended for use with :py:func:`scipp.signal.sosfiltfilt`. See there for
+    an example.
+
+    This is a wrapper around :py:func:`scipy.signal.butter`. See there for a
+    complete description of parameters. The differences are:
+
+    - Instead of a sampling frequency fs, this wrapper takes a data array ``da`` and a
+      dimension label ``dim`` as input. The sampling frequency is then computed from
+      ``da.coords[dim]``. Only inputs with regular sampling are supported.
+    - The critical frequency or frequencies must be provided as a variable with correct
+      unit.
+    - Only 'sos' output is supported.
+
+    :seealso: :py:func:`scipp.signal.sosfiltfilt`
+    """
+    coord = da.coords[dim]
+    fs = _frequency(coord).value
+    if not Wn.unit == one / coord.unit:
+        raise UnitError(
+            f"Critical frequency unit '{Wn.unit}' does not match sampling unit "
+            f"'{one / coord.unit}'")
+    import scipy.signal
+    return SOS(coord=coord.copy(),
+               sos=scipy.signal.butter(N=N, Wn=Wn.values, fs=fs, output='sos',
+                                       **kwargs))
+
+
+@wrap1d(keep_coords=True)
+def sosfiltfilt(da: DataArray, dim: str, *, sos: SOS, **kwargs) -> DataArray:
+    """
+    A forward-backward digital filter using cascaded second-order sections.
+
+    This is a wrapper around :py:func:`scipy.signal.sosfiltfilt`. See there for a
+    complete description of parameters. The differences are:
+
+    - Instead of an array ``x`` and an optional axis, the input must be a data array
+      and dimension label.
+    - The array of second-order filter coefficients ``sos`` includes the coordinate
+      used for computing the frequence used for creating the coefficients. This is
+      compared to the corresponding coordinate of the input data array to ensure that
+      compatible coefficients are used.
+
+    Examples:
+
+      >>> from scipp.signal import butter, sosfiltfilt
+      >>> x = sc.linspace(dim='x', start=1.1, stop=4.0, num=1000, unit='m')
+      >>> y = sc.sin(x * sc.scalar(1.0, unit='rad/m'))
+      >>> y += sc.sin(x * sc.scalar(400.0, unit='rad/m'))
+      >>> da = sc.DataArray(data=y, coords={'x': x})
+      >>> sos = butter(da, 'x', N=4, Wn=20 / x.unit)
+      >>> out = sosfiltfilt(da, 'x', sos=sos)
+    """
+    if not identical(da.coords[dim], sos.coord):
+        raise CoordError(f"Coord\n{da.coords[dim]}\nof filter dimension '{dim}' does "
+                         f"not match coord\n{sos.coord}\nused for creating the "
+                         "second-order sections representation by scipp.signal.butter.")
+    import scipy.signal
+    data = array(dims=da.dims,
+                 unit=da.unit,
+                 values=scipy.signal.sosfiltfilt(sos.sos, da.values, **kwargs))
```

Comment:
Not quite in the spirit of this review, but did you test this with multi-dim data?

Reply:
Yes, but only that it works (and output has correct properties), did not verify the result particularly for that case.

Reply:
Could you add a test, please? Both with smoothing in an inner and outer dim. I seem to remember having problems getting it work at some point.

Reply:
See update.